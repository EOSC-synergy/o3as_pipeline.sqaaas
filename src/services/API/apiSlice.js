import {createSlice, createAsyncThunk, createAction} from "@reduxjs/toolkit";
import {getModels, getPlotTypes, getPlotData} from "./client";
import {preTransformApiData} from "../../utils/optionsFormatter/optionsFormatter";
import {START_YEAR, END_YEAR} from "../../utils/constants";
import {setDisplayYRange, setDisplayYRangeForPlot} from "../../store/plotSlice/plotSlice";

/**
 * This object models an "enum" in JavaScript. Each of the values is used
 * to determine the state of an async request in the redux store. By accessing
 * the stored status components can render differently e.g. displaying a spinner,
 * displaying the fetched data on success or report an error.
 */
export const REQUEST_STATE = {
    idle: "idle",
    loading: "loading",
    success: "success",
    error: "error",
};

/**
 * This thunk action creater is created via redux toolkit. It allows to dispatch
 * asynchronous requests to the store. Internally a start action is dispatched
 * to the store that contains some information e.g. the request is currently loading
 * after the async request has been resolved it either returns the data or the error message
 *
 * This action creater is dispatched against the store at the beginning of the app
 * to fetch the models from the api.
 */
export const fetchModels = createAsyncThunk('api/fetchModels', async () => {
    const response = await getModels();
    return response.data;
});

/**
 * The description of fetchModels applies to this thunk action creater as well.
 *
 * This action creater is dispatched against the store at the beginning of the app
 * to fetch the plot types from the api.
 */
export const fetchPlotTypes = createAsyncThunk('api/fetchPlotTypes', async () => {
    const response = await getPlotTypes();
    return response.data;
});

/**
 * This function concatenates all given information into a string serving as an identifier
 * for cached requests in the store.
 *
 * @param {int} latMin specifies the minimum latitude
 * @param {int} latMax specifies the maximum latitude
 * @param {Array.<int>} months represents the selected months
 * @param {string} refModel the reference model to "normalize the data"
 * @param {int} refYear the reference year to "normalize the data"
 * @returns the generated string
 */
export const generateCacheKey = ({latMin, latMax, months, refModel, refYear}) => {
    return `lat_min=${latMin}&lat_min=${latMax}&months=${months.join(',')}&ref_meas=${refModel}&ref_year=${refYear}`;
}

/**
 * This action creator generates an action that is dispatched against the store
 * when the request of the data is started. The payload contains the plotId and the cacheKey.
 */
const fetchPlotDataPending = createAction("api/fetchPlotData/pending");
/**
 * This action creator generates an action that is dispatched against the store
 * when the request of the data succeded. The payload object contains the data, the cacheKey and
 * the plotId.
 */
const fetchPlotDataSuccess = createAction("api/fetchPlotData/success");
/**
 * This action creator generates an action that is dispatched against the store
 * when the request of the data failed. The playload object contains the
 * error message, the cacheKey and the plotId.
 */
const fetchPlotDataRejected = createAction("api/fetchPlotData/rejected");
/**
 * This action creator generates an action that is dispatched against the store
 * when the requested data is already in the cache and only needs to be selected.
 * The payload contains the plotId and the cacheKey.
 */
const selectExistingPlotData = createAction("api/selectPlotData");


export const updateDataAndDisplaySuggestions = ({plotId, cacheKey, data, modelsSlice}) => {

    return (dispatch, getState) => {
        dispatch(fetchPlotDataSuccess({data, plotId, cacheKey, modelsSlice}))
        const {min, max} = getState().api.plotSpecific[plotId].cachedRequests[cacheKey].suggested; // suggested min/max is availabe
        dispatch(setDisplayYRangeForPlot({plotId, minY: Math.floor(min / 10) * 10, maxY: Math.ceil(max / 10) * 10}));
    }
}

/**
 * This async thunk creator allows to generate a data fetching action that can be dispatched
 * against the store to start fetching new plot data from the api.
 *
 * @param {number} modelListBegin for faster testing limit fetching of model list
 * @param {number} modelListEnd for faster testing limit fetching of model list
 * @returns the async thunk action
 */
export const fetchPlotData = ({plotId, modelListBegin, modelListEnd}) => {

    return (dispatch, getState) => {
        if (typeof plotId === "undefined") plotId = getState().plot.plotId; // backwards compatible
        const latMin = getState().plot.generalSettings.location.minLat;
        const latMax = getState().plot.generalSettings.location.maxLat;
        const months = getState().plot.generalSettings.months;
        let modelList = getState().api.models.data.filter(modelName => modelName.includes("refC2") || modelName.includes("ACCESS"));
        /*
        if (typeof modelListBegin !== 'undefined' && typeof modelListEnd !== 'undefined') {
            modelList = modelList.slice(modelListBegin, modelListEnd);
        }
        */
        const startYear = START_YEAR;
        const endYear = END_YEAR;
        const refModel = getState().reference.settings.model;
        const refYear = getState().reference.settings.year;

        const cacheKey = generateCacheKey({latMin, latMax, months, refModel, refYear});
        // it shouldn't reload the same request if the data is already present (previous successful request) or 
        // if the request is already loading
        const cachedRequest = getState().api.plotSpecific[plotId].cachedRequests[cacheKey];

        if (typeof cachedRequest === "undefined") {
            // do nothing
        } else if (cachedRequest.status === REQUEST_STATE.loading) {
            // just select the current data obj (=> display loading spinner accordingly)
            dispatch(selectExistingPlotData({plotId, cacheKey}));
            return;
        } else if (cachedRequest.status === REQUEST_STATE.success) {
            dispatch(selectExistingPlotData({plotId, cacheKey}));
            const {min, max} = getState().api.plotSpecific[plotId].cachedRequests[cacheKey].suggested; // suggested min/max is availabe
            dispatch(setDisplayYRangeForPlot({plotId, minY: Math.floor(min), maxY: Math.floor(max)}));
            return Promise.resolve(); // request is already satisfied
        }

        // Initial action dispatched
        dispatch(fetchPlotDataPending({plotId, cacheKey}));

        // Return promise with success and failure actions

        return getPlotData({plotId, latMin, latMax, months, modelList, startYear, endYear, refModel, refYear})
            .then(
                response => dispatch(updateDataAndDisplaySuggestions({
                    plotId,
                    cacheKey,
                    data: response.data,
                    modelsSlice: getState().models
                })),
                error => dispatch(fetchPlotDataRejected({error: error.message, plotId, cacheKey})),
            );

    };
};

/**
 * This initial state describes how the data fetched from the api is stored in this
 * slice of the redux store.
 *
 * It should be noted that each additional data piece (models/plotTypes) follows the
 * same structure which is very useful to indicate the current status and store
 * potential errors or the returned data.
 */
const initialState = {
    models: {
        status: REQUEST_STATE.idle,
        error: null,
        data: [],
    },
    plotTypes: {
        status: REQUEST_STATE.idle,
        error: null,
        data: [],
    },
    plotSpecific: {
        tco3_zm: {
            active: null,
            cachedRequests: { // we need to cache: min,max, months, reference year and model

            }
        },
        tco3_return: {
            active: null,
            cachedRequests: {}
        },
    },
};

/**
 * The apiSlice is generated by the redux toolkit. This piece of the
 * store is responsible for storing all data fetched from the api.
 */
const apiSlice = createSlice({
    name: "api",
    initialState,
    reducers: {},
    /**
     * This interface is used to connect the redux thunk action which are returned
     * from the defined thunk action creators (fetchModels, fetchPlotTypes) to
     * this slice of the store.
     *
     * @param {object} builder an object handed by the redux toolkit to add these
     *                         "external" reducers to this slice
     */
    extraReducers(builder) {
        builder
            // fetch models
            .addCase(fetchModels.pending, (state) => {
                state.models.status = REQUEST_STATE.loading;
            })
            .addCase(fetchModels.fulfilled, (state, action) => {
                state.models.status = REQUEST_STATE.success;
                state.models.data = action.payload;
            })
            .addCase(fetchModels.rejected, (state, action) => {
                state.models.status = REQUEST_STATE.error;
                state.models.error = action.error.message;
            })

            // fetch plotTypes
            .addCase(fetchPlotTypes.pending, (state) => {
                state.plotTypes.status = REQUEST_STATE.loading;
            })
            .addCase(fetchPlotTypes.fulfilled, (state, action) => {
                state.plotTypes.status = REQUEST_STATE.success;
                state.plotTypes.data = action.payload;
            })
            .addCase(fetchPlotTypes.rejected, (state, action) => {
                state.plotTypes.status = REQUEST_STATE.error;
                state.plotTypes.error = action.error.message;
            })

            // fetch plot data
            .addCase(fetchPlotDataPending, (state, action) => {
                const {plotId, cacheKey} = action.payload;
                const plotSpecificSection = state.plotSpecific[plotId];

                plotSpecificSection.cachedRequests[cacheKey] = {
                    status: REQUEST_STATE.loading,
                    error: null,
                    data: [],
                    suggested: null, // this holds the suggested min / max values
                };
                plotSpecificSection.active = cacheKey; // select this request after dispatching it
            })
            .addCase(fetchPlotDataSuccess, (state, action) => {
                const {data, plotId, cacheKey, modelsSlice} = action.payload;
                const storage = state.plotSpecific[plotId].cachedRequests[cacheKey];
                storage.status = REQUEST_STATE.success;
                const {lookUpTable, min, max} = preTransformApiData({plotId, data, modelsSlice});
                storage.data = lookUpTable;
                storage.suggested = {min, max};

            })
            .addCase(fetchPlotDataRejected, (state, action) => {
                const {error, plotId, cacheKey} = action.payload;
                const storage = state.plotSpecific[plotId].cachedRequests[cacheKey];
                storage.status = REQUEST_STATE.error;
                storage.error = error;
            })

            // select existing data from cache
            .addCase(selectExistingPlotData, (state, action) => {
                const {plotId, cacheKey} = action.payload;
                state.plotSpecific[plotId].active = cacheKey; // update by chaning the active value to the existing cachekey
            })
    },
});

/**
 * The reducer combining all reducers defined in the plot slice.
 * This has to be included in the redux store, otherwise dispatching
 * the above defined actions wouldn't trigger state updates.
 */
export default apiSlice.reducer;

/**
 * This selectors allows components (=> the graph) to "listen" for the active plot data
 * of the current selected plot. That means whenever this data is about to change, e.g. because
 * a new fetch request is dispatched, the component rerenders.
 *
 * @param {object} state the state, handed by redux
 * @param {string} plotId identifies the plot
 * @returns the current active data for the active plot
 */
export const selectActivePlotData = (state, plotId) => {
    const plotSpecificSection = state.api.plotSpecific[plotId];
    const activeCacheKey = plotSpecificSection.active;
    if (activeCacheKey === null) { // no data is present yet, return dummy to indicate loading!
        return {
            status: REQUEST_STATE.loading,
        };
    }

    return plotSpecificSection.cachedRequests[activeCacheKey];
}