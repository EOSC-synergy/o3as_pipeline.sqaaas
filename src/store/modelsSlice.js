import { createSlice } from "@reduxjs/toolkit";

const STATISTICAL_VALUES = ["mean", "median", "derivative", "percentile"];

const initialState = {
    // currently active plot
    plotId: "tco3_zm",
    settings: {
        "tco3_zm": { 
            // this objects holds key-value-pairs, the keys being the model-group 
            // identifier and the values being the settings for each group 
            "all": { 
                // model group storing all information until it is possible 
                // to implement more model groups
                name: "All OCTS models",
                modelList: ["CCMI-1_ACCESS_ACCESS-CCM-refC2"],
                models: { // models is lookup table
                    "CCMI-1_ACCESS_ACCESS-CCM-refC2": { // single model
                        institute: "IMK",
                        dataset: {
                            x: [42],
                            y: [42],
                        },
                        color: "#000000",
                        plotStyle: "solid",
                        isVisible: true, // show/hide individual models from a group
                        mean: true,
                        derivative: true,
                        median: true,
                        percentile: true,
                    }
                },
                hidden: false, // show/hide complete group
                visibileSV: { // lookup table so the reducer impl. can be more convenient
                    mean: true,
                    derivative: true,
                    median: true,
                    percentile: true,
                }
            }
        },
        "tco3_return": {
            "all": { 
                name: "All Return/Recovery models",
                modelList: ["CCMI-1_ACCESS_ACCESS-CCM-refC2"],
                models: { // models is lookup table
                    "CCMI-1_ACCESS_ACCESS-CCM-refC2": { // single model
                        institute: "IMK",
                        dataset: {
                            x: [42],
                            y: [42],
                        },
                        hidden: false, // show/hide individual models from a group
                        mean: true,
                        derivative: true,
                        median: true,
                        percentile: true,
                        color: "#000000",
                        plotStyle: "solid",
                    }
                },
                hidden: false, // show/hide complete group
                visibileSV: { // lookup table so the reducer impl. can be more convenient
                    mean: true,
                    derivative: true,
                    median: true,
                    percentile: true,
                }
            }
        }
    }
}

const modelsSlice = createSlice({
    name: "models",
    initialState,
    reducers: {
        // set models
        // add models

        addModels(state, action) { 
            const { groupId, newModelList }  = action.payload;
            const activeSettings = state.settings[state.plotId];
            const selectedGroup = activeSettings[groupId];
            const activeModelList = selectedGroup.modelList;

            if (selectedGroup === undefined) {
                throw `Tried to access model-group with groupID "${groupId}" that is not yet defined!`;
            }

            // Add model to modelList if it is not already included
            for (var i = 0; i < newModelList.length; i++) {
                if (activeModelList.indexOf(newModelList[i]) !== -1) {
                    continue;
                }

                activeModelList.push(newModelList[i]);
            }

            // ---------------------------------------------------------
            // TODO: Add Model information to the model lookup table
            // ---------------------------------------------------------


        }, // expects groupID, list of models 
        removeModels(state, action) { 
            const { groupId, removeModelList }  = action.payload;
            const activeSettings = state.settings[state.plotId];
            const selectedGroup = activeSettings[groupId];
            const activeModelList = selectedGroup.modelList;

            if (selectedGroup === undefined) {
                throw `Tried to access model-group with groupID "${groupId}" that is not yet defined!`;
            }

            for (var i = 0; i < removeModelList.length; i++) {
                // Remove model from modelList
                activeModelList = activeModelList.filter(item => item !== removeModelList[i]);

                // Delete model object from models
                delete selectedGroup.models[removeModelList[i]];
            }

        }, // expects groupdID, list of models
        updatedModelGroup(state, action) { }, // implementing this requires more knowledge about the UI
        addedModelGroup(state, action) { },
        setVisibility(state, action) { 
            const { groupID, modelID, isVisible } = action.payload;
            const activeSettings = state.settings[state.plotId];
            const activeModel = activeSettings[groupID].models[modelID];
            if (activeModel === undefined) {
                throw `tried to access model with modelID "${modelID}" that is not yet defined!`;
            } 
            activeModel.isVisible = isVisible;
        },
        setStatisticalValuesIncluded(state, action) { // this is for an individual model
            // svType has to be a string of the form: 
            // mean | derivative | median | percentile
            const { groupID, modelID, svType, isIncluded } = action.payload;
            const activeSettings = state.settings[state.plotId];
            const activeModel = activeSettings[groupID].models[modelID];
            if (activeModel === undefined) {
                throw `tried to access model with modelID "${modelID}" that is not yet defined!`;
            }
            if (STATISTICAL_VALUES.includes(svType)) { // svType doesn't represent a valid statistical value
                throw `tried to set statistial value "${svType}" that is not a valid statistical value (${STATISTICAL_VALUES.join("|")})`;
            }
            activeModel[svType] = isIncluded;
        },
        setStatisticalValueForGroup(state, action) { // this is for an entire group
            const { groupID, svType, isIncluded } = action.payload;
            const activeSettings = state.settings[state.plotId];
            if (STATISTICAL_VALUES.includes(svType)) { // svType doesn't represent a valid statistical value
                throw `tried to set statistial value "${svType}" that is not a valid statistical value (${STATISTICAL_VALUES.join("|")})`;
            }
            activeSettings[groupID].visibileSV[svType] = isIncluded;
        },
    }
})

export const { 
    addModels,
    removeModels,
    updatedModelGroup,
    addedModelGroup,
    setVisibility, 
    setStatisticalValuesIncluded, 
    setStatisticalValueForGroup
} = modelsSlice.actions;

export default modelsSlice.reducer;

export const selectCurrentModelGroups = state => state.models[state.plot.plotId];
export const selectCurrentModelGroup = (state, groupId) => state.models[state.plot.plotId][groupId];