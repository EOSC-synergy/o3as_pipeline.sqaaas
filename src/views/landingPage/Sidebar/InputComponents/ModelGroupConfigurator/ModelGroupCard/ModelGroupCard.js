import * as React from 'react';
import Card from '@mui/material/Card';
import Typography from '@mui/material/Typography';
import VisibilityOffIcon from '@mui/icons-material/VisibilityOff';
import MuiVisibilityIcon from '@mui/icons-material/Visibility';
import { Checkbox, Divider, IconButton, FormControlLabel } from '@mui/material';
import Edit from '@mui/icons-material/Edit';
import Grid from '@mui/material/Grid';
import EditModelGroupModal from "../EditModelGroupModal/EditModelGroupModal";
import PropTypes from 'prop-types';
import { useSelector } from "react-redux";
import { useDispatch } from "react-redux"
import {
    STATISTICAL_VALUES,
    setStatisticalValueForGroup,
    selectStatisticalValueSettingsOfGroup,
    selectNameOfGroup,
    setVisibilityForGroup,
    selectVisibilityOfGroup
} from "../../../../../../store/modelsSlice/modelsSlice";

/**
 * a card containing information about the a modal group
 * @param {OBject} props 
 * @param {String} props.reportError - error function
 * @param {int} props.modelGroupId -> id of the model group
 * @returns {JSX.Element} a jsx containing a modal with a data grid with all models from the model group
 */
function ModelGroupCard(props) {
    
    const dispatch = useDispatch();
    const modelGroupName = useSelector(state => selectNameOfGroup(state, props.modelGroupId));
    const modelGroupStatisticalValue = useSelector(state => selectStatisticalValueSettingsOfGroup(state, props.modelGroupId));
    const isModelGroupVisible = useSelector(state => selectVisibilityOfGroup(state, props.modelGroupId));
    
    /**
     * toggles the visibility of the whole group in the graph
     * @todo insert redux call here
     */
    const toggleModelGroupVisibility = () => {
        dispatch(setVisibilityForGroup({groupId: props.modelGroupId, isVisible: !isModelGroupVisible}));
    }

    /**
     * toggles the visibility of the median from this group in the graph
     * @todo insert redux call here
     */
    const toggleModelGroupStatisticalValueVisibility = (event, statisticalValue) => {
        dispatch(setStatisticalValueForGroup(
            {groupId: props.modelGroupId, svType: statisticalValue, isIncluded: event.target.checked}
        ));
    }

    /**
     * state to keep track of whether the edit group modal is currently visible or not
     */
    const [isEditModalVisible, setEditModalVisible] = React.useState(false);

    /**
     * shows the edit group modal
     */
    const showEditModal = () => {
        setEditModalVisible(true);
    }

    /**
     * closes the edit modal
     */   
    const closeEditModal = () => {
        setEditModalVisible(false);
    }

    /**
     * function to return the visibility icon, depending on whether the group is currently visible or not
     */
    const VisibilityIcon = () => {
        return isModelGroupVisible ? <MuiVisibilityIcon /> : <VisibilityOffIcon />
    }

    return (
        <Card style={{margin: "5%"}} elevation={2}>
            <EditModelGroupModal id={props.modelGroupId} isOpen={isEditModalVisible} onClose={closeEditModal} />
            <Grid container>
                <Grid item xs={2}>
                    <IconButton aria-label="change visibility" onClick={toggleModelGroupVisibility}><VisibilityIcon /></IconButton>
                </Grid>
                <Grid item xs={8} textAlign="center">
                    <Typography variant="h6">{modelGroupName}</Typography>
                </Grid>
                <Grid item xs={2}>
                    <IconButton aria-label="edit" onClick={showEditModal}><Edit /></IconButton>
                </Grid>
            </Grid>
            <Divider />
            <Grid container sx={{paddingTop: '0.5em'}}>
                {Object.keys(STATISTICAL_VALUES).map((key, idx) => {
                    return (
                        <Grid item key={idx} xs={6} sx={{paddingLeft: '1em'}}>
                            <FormControlLabel
                                label={key}
                                id={`ModelGroupCoard-toggle-${key}-label`}
                                control={
                                    <Checkbox
                                        onChange={
                                            event => toggleModelGroupStatisticalValueVisibility(event, key)
                                        }
                                        checked={modelGroupStatisticalValue[key]}
                                        id={`ModelGroupCoard-toggle-${key}-checkbox`}
                                        labelid={`ModelGroupCoard-toggle-${key}-label`}
                                    />
                                }
                            />
                        </Grid>
                    );
                })}
            </Grid>
        </Card>
    )
}

ModelGroupCard.propTypes = {
    modelGroupId: PropTypes.number.isRequired,
    reportError: PropTypes.func
}

export default ModelGroupCard;