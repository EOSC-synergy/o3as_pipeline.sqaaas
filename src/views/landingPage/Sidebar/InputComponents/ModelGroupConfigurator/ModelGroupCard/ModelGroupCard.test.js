import { render, fireEvent, waitForElementToBeRemoved } from '@testing-library/react';
import '@testing-library/jest-dom';
import { createTestStore } from '../../../../../../store/store';
import ModelGroupCard, { getGroupName } from './ModelGroupCard';
import { Provider } from "react-redux";
import * as redux from 'react-redux';
import userEvent from '@testing-library/user-event';
import { REQUEST_STATE } from "../../../../../../services/API/apiSlice";
import { modelGroups } from "../../../../../../store/modelsSlice/modelsSlice";

let store;
let reportError;
const groupId = parseInt(Object.keys(modelGroups)[0]);
const groupName = modelGroups[groupId].name;
const visibileSV = modelGroups[groupId].visibileSV;
console.log(Object.keys(visibileSV));
const isVisble = modelGroups[groupId].isVisible;
describe('test ModelGroupCard rendering', () => {

    beforeEach(() => {
        store = createTestStore();
        reportError = jest.fn();
    });
    
    it('renders without crashing', () => {
        render(
            <Provider store={store}>
                <ModelGroupCard reportError={reportError} modelGroupId={groupId} />
            </Provider>
        );
    });

    it('renders correctly', () => {
        const { container } = render(
            <Provider store={store}>
                <ModelGroupCard reportError={reportError} modelGroupId={groupId} />
            </Provider>
        );
        expect(container).toMatchSnapshot();
    });

    it('renders modelGroup name', () => {
        const { getByTestId } = render(
            <Provider store={store}>
                <ModelGroupCard reportError={reportError} modelGroupId={groupId} />
            </Provider>
        );        
        expect(getByTestId(/groupName/)).toHaveTextContent(groupName);
    });

    test.todo('raises an error function if props.modelGroupId is not provided');

    it('raises a console.error function if props.reportError is not provided', () => {
        console.error = jest.fn();
        render(            
            <Provider store={store}>
                <ModelGroupCard modelGroupId={groupId} />
            </Provider>
        );
        expect(console.error).toHaveBeenCalled();
    });

    it('renders checkboxes correctly checked', () => {
        const { getByLabelText } = render(            
            <Provider store={store}>
                <ModelGroupCard modelGroupId={groupId} reportError={reportError} />
            </Provider>
        );

        for (let key in visibileSV) {
            expect(getByLabelText(key)).toBeInTheDocument();
            expect(getByLabelText(key)).toHaveProperty('checked', visibileSV[key]);
        }
    });

    it('renders visibility icon correctly', () => {
        const { getByTestId } = render(            
            <Provider store={store}>
                <ModelGroupCard modelGroupId={groupId} reportError={reportError} />
            </Provider>
        );
        expect(getByTestId(/VisibilityIcon-visible/)).toBeInTheDocument();
    });
});

describe('test ModelGroupCard functionality', () => {

    beforeEach(() => {
        store = createTestStore();
    });

    test.todo("dispatches setStatisticalValueForGroup with correct payload when checkbox is clicked");
    test.todo("dispatches setVisibilityForGroup with correct payload when icon is clicked");
    test.todo('test whether edit model group modal opens');
    test.todo('test whether add model group modal opens');

});
