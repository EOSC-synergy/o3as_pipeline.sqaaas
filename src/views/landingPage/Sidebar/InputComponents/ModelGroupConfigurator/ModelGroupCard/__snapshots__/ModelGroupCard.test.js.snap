// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ModelGroupCard rendering renders correctly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 MuiCard-root css-935hc6-MuiPaper-root-MuiCard-root"
    style="margin: 5%;"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 css-1o7apob-MuiGrid-root"
      >
        <button
          aria-label="change visibility"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ModelGroupCard-VisibilityIcon-visible"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8 css-epm2de-MuiGrid-root"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
        >
          blob
        </h6>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2 css-1o7apob-MuiGrid-root"
      >
        <button
          aria-label="edit"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="EditIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <div
      class="MuiGrid-root MuiGrid-container css-m6sf28-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-d83jk5-MuiGrid-root"
      >
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
          id="ModelGroupCoard-toggle-mean-label"
        >
          <span
            class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
            labelid="ModelGroupCoard-toggle-mean-label"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              id="ModelGroupCoard-toggle-mean-checkbox"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            mean
          </span>
        </label>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-d83jk5-MuiGrid-root"
      >
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
          id="ModelGroupCoard-toggle-median-label"
        >
          <span
            class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
            labelid="ModelGroupCoard-toggle-median-label"
          >
            <input
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              id="ModelGroupCoard-toggle-median-checkbox"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxOutlineBlankIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            median
          </span>
        </label>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-d83jk5-MuiGrid-root"
      >
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
          id="ModelGroupCoard-toggle-derivative-label"
        >
          <span
            class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
            labelid="ModelGroupCoard-toggle-derivative-label"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              id="ModelGroupCoard-toggle-derivative-checkbox"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            derivative
          </span>
        </label>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-d83jk5-MuiGrid-root"
      >
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
          id="ModelGroupCoard-toggle-percentile-label"
        >
          <span
            class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
            labelid="ModelGroupCoard-toggle-percentile-label"
          >
            <input
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              id="ModelGroupCoard-toggle-percentile-checkbox"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CheckBoxOutlineBlankIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            percentile
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;
